### ISSUE TITLE
Lexer cannot process UTF-8 files with a byte order mark (BOM).

### ISSUE THREAD

**Issue by ManuelAtWork** [mesonbuild/meson#8413]
**Description**
When meson.build files (or cross files) start with the utf-8 BOM (sequence 0xEF 0xBB 0xBF), then Meson's lexer fails, e.g.:

    ..\subprojects\poco\meson.build:1:0: ERROR: lexer
    ﻿project('poco', 'cpp')

**Reproduce**

1. open meson.build in VIM
2. `:set bomb`
3. save & exit
4. run `meson setup builddir`

**Expected behavior**
Meson's lexer peeks the first three bytes of the input file. If they match the BOM, then read & discard. (And be happy to have a UTF-8 file.)

**system parameters**
* `meson --version` yields `0.57.1`

**Comment by jpakkane** [comment]
There is no reason to ever use BOMs in Meson files. Any editor that produces them by default is doing it wrong.

**Comment by ManuelAtWork** [comment]
I agree that it is not strictly necessary, but tolerating a BOM would improve user experience for a (presumambly) low effort.

If a `meson.build` does (errnously) have a BOM, then it does not emit a proper error message. Instead it complains about the `project` in the first line, which is a bit misleading:

    ..\meson.build:1:0: ERROR: lexer
     project('...', 'cpp')
    ^

For less experienced developers this is a hard-to-diagnose error and they may spend significant time finding the root cause. (I argue that this behaviour sucks, which is a defect in a "non-sucking" build system. ;-) )

**Our use case**
We have something that could be called 'encoding hell': A huge, 40-year old code base with call kinds of latin, cyrillic, greek and line-drawing characters in different files. A few years ago we decided to convert everything utf-8 as we go, and converted files will be tagged with a BOM to safely distinguish them from the rest. That is why our editors create new files with a BOM automatically. This has been very effective and saved us countless working hours. 


I have attached a little C++ function that we use in our lexers to read&discard the BOM. It is called immediately after the file/stream has been opened. [ByteOrderMark.zip](https://github.com/mesonbuild/meson/files/6048225/ByteOrderMark.zip)

**Comment by denizzzka** [comment]
Years ago I came across same type of issue in D compiler or build tool - it was very difficult to understand where is problem really was, I spent several days. As for me, it is better is to display a more detailed error message like: "BOM found, please remove it"

**Comment by xclaesse** [comment]
As far as I understand supporting BOM is a simple matter of using `utf-8-sig` encoding in python when reading the file. I don't see any downside of doing that.

That being said, I came across that `ERROR: lexer` and it's a really bad error message, we should change that.

**Comment by jpakkane** [comment]
I checked and this is what the official Unicode standards document says about the issue:

> Use of a BOM is neither required nor recommended for UTF-8, but may be encountered in contexts where UTF-8 data is converted from other encoding forms that use a BOM or where the BOM is used as a UTF-8 signature

In this case we know that the files are never converted, because Meson build files have always been utf-8.
